hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/StorageBasedLockProvider.java

```diff
@@ -0,0 +1,531 @@
+    switch (result.getLeft()) {
+      case UNKNOWN_ERROR:
+        // Here we do not know the state of the lock.
+        logErrorLockState(FAILED_TO_RELEASE, "Lock state is unknown.");
+        return false;
+      case SUCCESS:
+        logInfoLockState(RELEASED);
+        setLock(null);
+        return true;
+      case ACQUIRED_BY_OTHERS:
+        // As we are confident no lock is held by itself, clean up the cached lock object.
+        // However this is an edge case, so warn.
+        logWarnLockState(RELEASED, "lock should not have been acquired by others.");
+        setLock(null);
+        return true;
+      default:
+        throw new HoodieLockException("Unexpected lock update result: " + result.getLeft());
+    }
+  }
+
+  /**
+   * Renews (heartbeats) the current lock if we are the holder, it forcefully set
+   * the expiration flag
+   * to false and the lock expiration time to a later time in the future.
+   * @return True if we successfully renewed the lock, false if not.
+   */
+  @VisibleForTesting
+  protected synchronized boolean renewLock() {
+    try {
+      // If we don't hold the lock, no-op.
+      if (!believesLockMightBeHeld()) {
+        logger.warn("Owner {}: Cannot renew, no lock held by this process", ownerId);
+        // No need to extend lock lease.
+        return false;
+      }
+
+      long oldExpiration = getLock().getValidUntil();
+      // Attempt conditional update, extend lock. There are 3 cases:
+      // 1. Happy case: lock has not expired yet, we extend the lease to a longer
+      // period.
+      // 2. Corner case 1: lock is expired and is acquired by others, lock renewal
+      // failed with ACQUIRED_BY_OTHERS.
+      // 3. Corner case 2: lock is expired but no one has acquired it yet, lock
+      // renewal "revived" the expired lock.
+      // Please note we expect the corner cases almost never happens.
+      // Action taken for corner case 2 is just a best effort mitigation. At least it
+      // prevents further data corruption by
+      // letting someone else acquire the lock.
+      Pair<LockUpdateResult, StorageLockFile> currentLock = this.lockService.tryCreateOrUpdateLockFileWithRetry(
+          () -> new StorageLockData(false, System.currentTimeMillis() + validitySeconds, ownerId),
+          getLock(),
+          LOCK_UPSERT_RETRY_COUNT);
+      switch (currentLock.getLeft()) {
+        case ACQUIRED_BY_OTHERS:
+          logger.error("Owner {}: Unable to renew lock as it is acquired by others.", ownerId);
+          // No need to extend lock lease anymore.
+          return false;
+        case UNKNOWN_ERROR:
+          // This could be transient, but unclear, we will let the heartbeat continue
+          // normally.
+          // If the next heartbeat run identifies our lock has expired we will error out.
+          logger.warn("Owner {}: Unable to renew lock due to unknown error, could be transient.", ownerId);
+          // Let heartbeat retry later.
+          return true;
```

Comment:
Should this return `false` based on the docs?

Reply:
Docs should be updated. We should always retry renewal unless we receive a known fatal error code.